crates/pathfinder/src/storage/schema/revision_0018.rs

```diff
@@ -0,0 +1,202 @@
+///
+/// This is achieved by re-creating the `starknet_blocks` table, shifting the PK
+/// from `number` to `hash`. This in turn requires re-creating any tables with a FK
+/// on the original `starknet_blocks` table.
+///
+/// In addition, a `canonical_blocks` table is created to track the canonical block chain
+/// now that forks are allowed in `starknet_blocks`.
+pub(crate) fn migrate(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute(
+        r"-- Stores StarkNet block headers.
+CREATE TABLE starknet_blocks_new (
+    hash      BLOB    PRIMARY KEY NOT NULL,
+    number    INTEGER NOT NULL,
+    root      BLOB    NOT NULL,
+    timestamp INTEGER NOT NULL, 
+    gas_price BLOB    NOT NULL,
+    sequencer_address BLOB NOT NULL,
+    version_id INTEGER REFERENCES starknet_versions(id)
+)",
+        [],
+    )
+    .context("Creating new starknet_blocks table")?;
+
+    tx.execute(
+        "INSERT INTO starknet_blocks_new(hash,number,root,timestamp,gas_price) SELECT hash,number,root,timestamp,gas_price FROM starknet_blocks",
+        [],
+    )
+    .context("Copying starknet_blocks data")?;
+
+    create_canonical_chain(tx).context("Creating canonical_blocks table")?;
+    migrate_state_updates(tx).context("Migrating starknet_state_updates table")?;
+    migrate_events(tx).context("Migrating starknet_events table")?;
+
+    tx.execute("DROP TABLE starknet_blocks", [])
+        .context("Dropping old starknet_blocks table")?;
+    tx.execute(
+        "ALTER TABLE starknet_blocks_new RENAME TO starknet_blocks",
+        [],
+    )
+    .context("Renaming new starknet_blocks table")?;
+
+    Ok(())
+}
+
+/// Re-creates the `starknet_state_updates_new` table, updating the `block_hash` FK to
+/// reference the new `starknet_blocks_new` table.
+fn migrate_state_updates(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute(
+        r"-- Stores StarkNet state updates. 
+CREATE TABLE starknet_state_updates_new (
+    block_hash BLOB PRIMARY KEY NOT NULL,
+    data BLOB NOT NULL,
+    FOREIGN KEY(block_hash) REFERENCES starknet_blocks_new(hash) ON DELETE CASCADE
+)",
+        [],
+    )
+    .context("Creating new table")?;
+
+    tx.execute(
+        "INSERT INTO starknet_state_updates_new(block_hash, data) SELECT block_hash, data FROM starknet_state_updates",
+        [],
+    )
+    .context("Copying data")?;
+
+    tx.execute("DROP TABLE starknet_state_updates", [])
+        .context("Dropping old table")?;
+    tx.execute(
+        "ALTER TABLE starknet_state_updates_new RENAME TO starknet_state_updates",
+        [],
+    )
+    .context("Renaming new table")?;
+
+    Ok(())
+}
+
+/// Re-creates the `starknet_events` events table, updating the `block_number` FK
+/// to reference the `canonical_blocks` table instead.
+fn migrate_events(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute(
+        r"CREATE TABLE starknet_events_new (
+    block_number  INTEGER NOT NULL,
```

Comment:
Thanks I had forgotten about this. I initially tried with `id` but the data copying did not work; but I've since made other fixes so maybe it will work now. Will try again.

Reply:
Done.